pkg/util/client/cache_reader.go

```diff
@@ -0,0 +1,263 @@
+/*
+Copyright 2022 The Kruise Authors.
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package client
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"strings"
+	"time"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	apimeta "k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/selection"
+	toolscache "k8s.io/client-go/tools/cache"
+	"k8s.io/klog/v2"
+	"sigs.k8s.io/controller-runtime/pkg/cache"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/client/apiutil"
+)
+
+var DisableDeepCopy = disableDeepCopy{}
+
+type disableDeepCopy struct{}
+
+func (_ disableDeepCopy) ApplyToList(_ *client.ListOptions) {
+}
+
+type cacheReader struct {
+	cache  cache.Cache
+	scheme *runtime.Scheme
+}
+
+func (r *cacheReader) Get(ctx context.Context, key client.ObjectKey, out client.Object) error {
+	gvk, err := apiutil.GVKForObject(out, r.scheme)
```

Comment:
is it neccessary to copy the CacheReader function here? delegate reader will not paas Get request to cacheReader, it seems this func is useless

Reply:
A Reader interface contains Get and List methods. The cacheReader here is going to be a complete Reader without any deepcopy, even if we just use the List for now.